cl_khr_tensor extension (#2)
pjaaskel/OpenCL-Docs: ext/cl_khr_tensor.asciidoc

// Copyright 2023 The Khronos Group. This work is licensed under a
// Creative Commons Attribution 4.0 International License; see
// http://creativecommons.org/licenses/by/4.0/
= cl_khr_tensor

:source-highlighter: coreray

[[cl_khr_tensor]]
== Tensor Data Type

> Purpose of this extension is to provide ...

Reviewer: "This extension provides a new opaque OpenCL datatype called `cl_tensor`. It is used for storing N-dimensional tensor data in implementation-optimized way. The datatype is designed to be efficiently used within the `cl_khr_command_buffers` extension to capture task graphs which can utilize tensors as input, output and temporary storage."  [etc etc]

Author: Added this and tweaked it a little.

=== General information

==== Name Strings

`cl_khr_tensor`

==== Version history

[cols="1,1,3",options="header",]
|====
| *Date*     | *Version* | *Description*
| 2023-10-XX | 0.1.0     | First assigned version.
|====

==== Dependencies

This extension is written against the OpenCL Specification version 3.0.14.

This extension requires OpenCL 1.2 or later.

> This extension requires cl_khr_command_buffer.

Reviewer: Does it require it or just refer to it and benefit from it?

Reviewer: Let's use cl_exp_tensor for the name to denote 'experimental' until it's approved as a 'khr' to avoid confusions.

Author: Made cl_khr_command_buffer optional, renamed the extension to cl_exp_tensor.

==== Contributors

Henry Linjamäki, Intel. +

> 

Reviewer: Please list me as well here. I believe I've contributed somewhat over our discussions.

Author: Done. Added @bashbaug too.

=== Overview

=== Modifications to OpenCL

==== New OpenCL Functions

To create a tensor use:

[source,c]
----
cl_tensor clCreateTensor(
    cl_context context,
    const cl_tensor_peoperties *properties,
    size_t rank,

>     size_t shape,

Reviewer 2: Given the description below I think this should be:

```suggestion
    const size_t* shape,
```

>     cl_tensor_type dtype,

Reviewer: should it be `cl_tensor_datatype`?

  special property 0. If no properties are required, properties may be
  NULL.

* _rank_ is the number of dimensions. Zero value creates a "scalar"
  tensor which has no dimensions but has storage for one element.

* _shape_ is a list of sizes of the dimensions. The length of the list
  must be _rank_ elements. _shape_ can be NULL if _rank_ value is
  zero. All the first _rank_ values in the list must be non-zero.

* _dtype_ is the element type of _tensor_. Refer to the
  <<TensorDtypes>> table for the types.

* _errcode_ret_ may return an appropriate error code. If errcode_ret
  is NULL, no error code is returned.

clCreateTensor function creates a `rank`-dimensional tensor with
`shape[0] * shape[1] * ... * shape[rank-1]` elements of _dtype_
type. At the creation time of the tensor, it does not have
storage. The storage is assigned to the tensor either by:

* calling clCreateBufferWithProperties() with CL_MEM_BIND_TO_TENSOR or

* automatically by command buffers - possibly on-demand basis - if the
  tensor is created with CL_TENSOR_COMMAND_BUFFER_TEMPORARY property
  set on.

A command that refers to a tensor must be bound to a valid buffer

> object before enqueuing the command into a command queue unless the

Reviewer: ```suggestion
object before enqueuing the command that refers to the tensor into a command queue unless the
```

command is recorded in a command buffer and
CL_TENSOR_COMMAND_BUFFER_TEMPORARY is set to true.

*clCreateTensor* returns a valid non-zero tensor object and errcode_ret
is set to CL_SUCCESS if the tensor object is created
successfully. Otherwise, they return a NULL value with one of the
following error values returned in errcode_ret:

* CL_INVALID_CONTEXT if context is not a valid context.

* CL_INVALID_PROPERTY if a property name in properties is not a
  supported property name, if the value specified for a supported
  property name is not valid, or if the same property name is
  specified more than once.

* CL_INVALID_VALUE if a value specified in dtype is invalid.

* CL_OUT_OF_HOST_MEMORY if there is a failure to allocate resources
  required by the OpenCL implementation on the host.

.Tensor element types
[cols="1,2",stripes=odd]
[#TensorDtypes]
|===
| *Tensor element data type* | *Description*

| CL_TENSOR_BOOL       | 1-bit signedless integer.
| CL_TENSOR_INT8       | 8-bit signed integer.
| CL_TENSOR_INT16      | 16-bit signed integer.
| CL_TENSOR_INT32      | 32-bit signed integer.
| CL_TENSOR_INT64      | 64-bit signed integer.

> | CL_TENSOR_UINT8      | 8-bit signed integer.

Reviewer: ```suggestion
| CL_TENSOR_UINT8      | 8-bit unsigned integer.
```
etc. also others miss 'un'

| CL_TENSOR_UINT16     | 16-bit signed integer.
| CL_TENSOR_UINT32     | 32-bit signed integer.
| CL_TENSOR_UINT64     | 64-bit signed integer.

> | CL_TENSOR_HALF       | Half precision floating-point value.

Reviewer: ```suggestion
| CL_TENSOR_HALF       | Half precision floating-point.
```
remove 'value' for consistency

| CL_TENSOR_BFLOAT16   | 16-bit brain floating-point value.
| CL_TENSOR_FLOAT      | Single precision floating-point value.
| CL_TENSOR_DOUBLE     | Double precision floating-point value.
| CL_TENSOR_COMPLEX64  | 64-bit complex floating point value with
  32-bit real and imaginary part.
| CL_TENSOR_COMPLEX128 | 128-bit complex floating point value with
  64-bit real and imaginary part.
|===

.Tensor properties
[cols="2,1,2",stripes=odd]
|===
| *Tensor Property* | *Property Value* | *Description*

| CL_TENSOR_COMMAND_BUFFER_TEMPORARY | cl_bool

a| If the value is true, create a "temporary" tensor that only can be
used on commands recorded in command buffers. The storage of the
temporary tensors are managed by command buffers. When a temporary

> tensor is used by multiple command buffer, the tensor receive separate

Reviewer: ```suggestion
tensor is used by multiple command buffers, the tensor receives separate
```

storage for each command buffer.

// IOW, Data may not be exchanged between command buffers through
// temporary tensors.

Temporary tensors may not be bound to buffer objects.

Data stored in temporary tensors are not preserved across command
buffer executions.
|===

To retain a tensor object, call the function

[source,c]
----

> cl_int clRetainTensorObject(

Reviewer: it should fit to a single line

  cl_tensor tensor);
----

* _tensor_ is the tensor object to be retained.

The _tensor_ reference count is incremented.

*clRetainTensor* returns CL_SUCCESS if the function is executed
successfully. Otherwise, it returns one of the following errors:

> * CL_INVALID_TENSOR if tensor is not a valid tensor object.

Reviewer: ```suggestion
* CL_INVALID_TENSOR if the tensor is not a valid tensor object.
```

To release a tensor object, call the function

[source,c]
----

> cl_int clReleaseTensorObject(

Reviewer: single line?

  supported param_name types and the information returned in
  _param_value_ by clGetTensorInfo is described in the <<Tensor Object
  Queries>> table.

* _param_value_ is a pointer to memory where the appropriate result
  being queried is returned. If _param_value_ is NULL, it is ignored.

* _param_value_size_ is used to specify the size in bytes of memory
  pointed to by _param_value_. This size must be ≥ size of return type
  as described in the <<Tensor Object Queries>> table.

* _param_value_size_ret_ returns the actual size in bytes of data
  being queried by _param_name_. If _param_value_size_ret_ is NULL, it is
  ignored.

*clGetTensorInfo* returns CL_SUCCESS if the function is executed
 succesfully. Otherwise, it returns one of the following errors:

* CL_INVALID_TENSOR if _tensor_ is not a valid tensor object.

[#Tensor Object Quaries]
.List of supported param_names by clGetTensorInfo
[cols="2,1,2",stripes=odd]
|===
| CL_TENSOR_RANK  | size_t         | Return the tensor rank.
| CL_TENSOR_SHAPE | size_t[]       | Return the tensor shape.
| CL_TENSOR_DTYPE | cl_tensor_type | Return the tensor data type.

| CL_TENSOR_COMMAND_BUFFER_TEMPORARY | cl_bool | Return true if the

> tensor is temporary tensor for command buffers.

Reviewer: ```suggestion
tensor is a temporary tensor for command buffers.
```

| CL_TENSOR_BOUND_TO_BUFFER | cl_bool | Return true if the tensor is
bound to a buffer. If CL_TENSOR_COMMAND_BUFFER_TEMPORARY is true, then
CL_TENSOR_BOUND_TO_BUFFER must return false.

| CL_TENSOR_BUFFER | cl_mem a| If CL_TENSOR_BOUND_TO_BUFFER is true,
return the buffer object the tensor is bound to. Otherwise,
clGetTensorInfo call returns:

* CL_INVALID_MEM_OBJECT if the tensor is not bound to a buffer object.

* CL_INVALID_PROPERTY otherwise.

| CL_TENSOR_CONTEXT | cl_context | Return the context specified when
  the tensor object is created.

| CL_TENSOR_REFERENCE_COUNT | cl_uint | Return the tensor reference
count.
|===

> To read from a tensor to host memory / buffer object or to write to a

Reviewer: ```suggestion
The following functions are for reading from a tensor to host memory / buffer object or to write to a
```

> tensor object from host memory / buffer object call one of the functions.

Reviewer: ```suggestion
tensor object from host memory / buffer object.
```

[source,c]
----

> cl_int clEnqueueReadTensor(

Reviewer 2: Presumably these read and write functions are invalid for `CL_TENSOR_COMMAND_BUFFER_TEMPORARY ` tensors?

Author: Yes and the functions should return `CL_INVALID_OPERATION`. This is attempted to be stated in `CL_MEM_COMMAND_BUFFER_TEMPORARY` property description (perhaps it's not clear enough?).

  cl_command_queue command_queue,
  cl_tensor tensor,
  cl_bool blocking_command,
  cl_mem buffer,
  void* host_ptr,
  cl_uint num_events_in_wait_list,
  const cl_event* event_wait_list,
  cl_event* event);
----

[source,c]
----
cl_int clEnqueueWriteTensor(
  cl_command_queue command_queue,
  cl_tensor tensor,
  cl_bool blocking_command,
  cl_mem buffer,

>   void* host_ptr,

Reviewer 2: This should be `const` for the write API.  Not sure if other changes are needed here too (possible copy-paste?).

```suggestion
  const void* host_ptr,
```

  cl_uint num_events_in_wait_list,
  const cl_event* event_wait_list,
  cl_event* event);
----

* _command_queue_ is a valid host command-queue in which the read /
  write command will be queued. _command_queue_ and _tensor_ must be
  created with the same OpenCL context.

* _tensor_ refers to a valid tensor object which is bound to a buffer.

* _blocking_command_ indicate if the read and write operations are
  blocking or non-blocking (see below).

* _buffer_ refers to a valid buffer object where data is to be
  read into or to be written from when the value of _host_ptr_ is

>   NULL. If _host_ptr_ is non-NULL then value of _buffer_ is ignored.

Reviewer 2: It might be less error-prone to require the buffer to be NULL when the host pointer is non-NULL.

Author: This should be now reflected in the extension.

  _event_wait_list_ is NULL, then this particular command does not
  wait on any event to complete. If _event_wait_list_ is NULL,
  _num_events_in_wait_list_ must be 0. If _event_wait_list_ is not
  NULL, the list of events pointed to by _event_wait_list_ must be
  valid and _num_events_in_wait_list_ must be greater than 0. The
  events specified in _event_wait_list_ act as synchronization
  points. The context associated with events in _event_wait_list_ and
  _command_queue_ must be the same. The memory associated with
  _event_wait_list_ can be reused or freed after the function returns.

* _event_ returns an event object that identifies this read / write
  command and can be used to query or queue a wait for this command to
  complete. If _event_ is NULL or the enqueue is unsuccessful, no
  event will be created and therefore it will not be possible to query
  the status of this command or to wait for this command to
  complete. If _event_wait_list_ and _event_ are not NULL, _event_
  must not refer to an element of the _event_wait_list_ array.

For a read and write operation, the elements of N-dimensional tensor are

> related to host memory / buffer object as followed:

Reviewer: ```suggestion
related to host memory / buffer object as follows:
```

----

> tensor.element(i0, i1, ..., i<N-2>, i<N-1>)) == (tensor.dtype)buffer_or_host_ptr[

Reviewer: ```suggestion
tensor.element(i0, i1, ..., i<N-2>, i<N-1>) == (tensor.dtype)buffer_or_host_ptr[
```

Reviewer: We discussed this aspect, but I forgot the outcome. Do we want to tie the memory layout or let it be implementation specific with implementation defining conversion calls to the common layout? Or did we decide to let the implementation optimize the layout optimally inside a captured command buffer?

  i0 * tensor.shape[1] * tensor.shape[2] * ... * tensor.shape[N-1] +
  i1 * tensor.shape[2] * tensor.shape[3] * ... * tensor.shape[N-1] +
  ... +
  i<N-2> * tensor.shape[i(N-1)] +
  i<N-1>]
----

Where `iX` is a tensor coordinate index with inclusive range of `0..<shape[X]>`.

> // TODO: add clEnqueueCopyTensor

Reviewer: Yes, makes sense to me as the 'metadata format' is implementation specific although the raw data content is not, right?

// TODO: add clEnqueueFillTensor?

// TODO: add command buffer variants for clEnqueue{copy,read,write}Tensor.

==== Add New Buffer Property in Section 5.2.1

[cols="2,1,2",stripes=odd]
|===
| CL_MEM_BIND_TO_TENSOR | cl_tensor a| Use the created buffer as
storage for the given valid tensor. To succeed creating the buffer,
the target tensor may not have storage already, must not have

> CL_TENSOR_COMMAND_BUFFER_TEMPORARY property set on and _size_ argument

Reviewer: ```suggestion
CL_TENSOR_COMMAND_BUFFER_TEMPORARY property set on and the _size_ argument
```

of the clCreateBufferWithProperties() must be zero.

Size of the memory buffer is implementation-defined and it can be
queried with clGetTensorInfo().

> Memory layout of the tensor in the created memory buffer is

Reviewer: Ah. Now I understood it. The above memory layout is only for the host in/out representation? That should be clarified above.

Author: Might be clarified now in the clEnqueueTranslate{To,From}Tensor description.

implementation-defined and opaque to the applications and it may
change at unspecified points. Implementation may store auxiliary data
in the memory buffer for the tensor. Therefore, writing data into the
memory buffer directly using the cl_mem handle leads to undefined
behavior.

If the tensor is already bound to a buffer object,
clCreateBufferWithProperties call returns CL_TENSOR_BOUND_TO_BUFFER
error code.
|===

=== Sample Codes

> Helper functions used in the follow up tensor code samples:

Reviewer: Let's first define a couple of helper functions which are used in the tensor usage examples:

[source,c]
----
cl_kernel create_matmul_kernel(

>   cl_context ctx, std::span<cl_device_id> device_span,

Reviewer: What's the `device_span` here? Shall we stick to C in the examples?

Author: Should be fine to use C++ in the examples. `cl_khr_command_buffer` uses C++ in their [example](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#_sample_code).

Reviewer: Still, the device_span seems to be not used?

Reviewer: (Well, maybe it is indeed in the 'omitted'-part.)

  cl_tensor lhs, cl_tensor rhs, cl_tensor out) {
  // A hypothetical matmul kernel signature in pseudo OpenCL C for
  // illustrative purposes:
  //
  //   kernel void matmul(
  //     global read_only tensor_t,
  //     global read_only tensor_t,
  //     global write_only tensor_t);

  cl_kernel matmul_kernel = /* Omitted. */;
  clSetKernelArg(matmul_kernel, 0, sizeof(cl_tensor), &lhs);
  clSetKernelArg(matmul_kernel, 1, sizeof(cl_tensor), &rhs);
  clSetKernelArg(matmul_kernel, 2, sizeof(cl_tensor), &out);
  return matmul_kernel;
}

> cl_kernel create_matmul_kernel(

Reviewer: ```suggestion
cl_kernel create_add_kernel(
```

  cl_context ctx, std::span<cl_device_id> device_span,
  cl_tensor lhs, cl_tensor rhs, cl_tensor out) {
  // A hypothetical add kernel signature in pseudo OpenCL C for illustrative
  // purposes:
  //
  // kernel void add(
  //     global read_only tensor_t,
  //     global read_only tensor_t,
  //     global write_only tensor_t);

  cl_tensor add_kernel = /* Omitted. */;
  clSetKernelArg(add_kernel, 0, sizeof(cl_tensor), &lhs);
  clSetKernelArg(add_kernel, 1, sizeof(cl_tensor), &rhs);
  clSetKernelArg(add_kernel, 2, sizeof(cl_tensor), &out);
  return add_kernel;
}
----
An example usage of tensors on a command queue:

[source,c]
----
constexpr size_t b = 64, m = 100, n = 200, k = 50;

> cl_tensor in0 = clCreateTensor(ctx, nullptr, 3, {b, m, k}, CL_TENSOR_FLOAT, err);

Reviewer: If we stick to C (like we should since the OpenCL spec is a C spec), we should use NULL instead of nullptr, right?

// Command buffers allocate space for them as needed.
//
// NOTE: same temporary tensor handle used in multiple command buffers
//       will have separate storage. IOW, command buffers may not exchange
//       data via temporary buffers between them.
cl_tensor in0 = clCreateTensor(ctx, {CL_TENSOR_COMMAND_BUFFER_TEMPORARY, true, 0},
  3, {b, m, k}, CL_TENSOR_FLOAT, err);
cl_tensor in1 = clCreateTensor(ctx, {CL_TENSOR_COMMAND_BUFFER_TEMPORARY, true, 0},
  3, {b, k, n}, CL_TENSOR_FLOAT, err);
cl_tensor in2 = clCreateTensor(ctx, {CL_TENSOR_COMMAND_BUFFER_TEMPORARY, true, 0},
  3, {b, m, n}, CL_TENSOR_FLOAT, err);
cl_tensor t0  = clCreateTensor(ctx, {CL_TENSOR_COMMAND_BUFFER_TEMPORARY, true, 0},
  3, {b, m, n}, CL_TENSOR_FLOAT, err);
cl_tensor out = clCreateTensor(ctx, {CL_TENSOR_COMMAND_BUFFER_TEMPORARY, true, 0},
  3, {b, m, n}, CL_TENSOR_FLOAT, err);

cl_kernel matmul_kernel = create_matmul_kernel(ctx, device_span, in0, in1, t0);
cl_kernel add_kernel = create_add_kernel(ctx, device_span, t0, in2, out);

// Binding a buffer to temporary tensor is not allowed.
auto ignored = clCreateBufferWithProperties(
  ctx, {CL_MEM_BIND_TO_TENSOR, t0, 0}, CL_MEM_READ_WRITE, 0, nullptr, &err);

> assert(err == CL_TENSOR_IS_TEMPORARY)

Reviewer: ```suggestion
assert(err == CL_TENSOR_IS_TEMPORARY);
```

clCommandWriteTensorKHR(
  cmd_b, cmd_q, in1, false, nullptr, nullptr, {b, k, m}, nullptr,
  in1_data.data(), 0, nullptr, &in1_syncp);
clCommandNDRangeKernelKHR(
  cmd_b, cmd_q, nullptr, matmul_kernel, 0, nullptr, nullptr, nullptr,
  2, {in0_syncp, in2_syncp}, &matmul_syncp, nullptr);
clCommandNDRangeKernelKHR(
  cmd_b, cmd_q, nullptr, add_kernel, 0, nullptr, nullptr, nullptr,
  1, {matmul_syncp}, &add_syncp, nullptr);
clCommandReadTensorKHR(
  cmd_b, cmd_q, out,  false, nullptr, nullptr, {b, k, m}, nullptr,
  out_data.data(), 1, {add_syncp}, nullptr);

// Finalize the command buffer. At this point the OpenCL
// implementation may reserve enough storage for all the tensor
// temporaries. Temporary tensors might be eliminated - for example,
// OpenCL implementation could use 'out' tensor to store result of
// matmul_kernel , thus, eliminating the need of 't0' tensor.
clFinalizeCommandBufferKHR(cmd_b);

// Temporary tensors used in a command buffer can't be read or written
// into. A hypothetical reason is that the finalized command buffer
// might not use some of the tensor.
assert(clEnqueueReadTensor(..., t0, ...) == CL_INVALID_OPERATION);
----

> 

Reviewer: The OpenCL C builtins for acessing the tensor metadata needs to be defined (and after that the SPIR-V extensions).

Author: Could the OpenCL C support for tensors be in a separate extension? I think, there is a lot to be covered. 

Built-in kernels should be able to define kernels tensor arguments without the OpenCL C tensor extension.

Reviewer: Yes. Since this will be useful already for the BiKs, it doesn't _require_ the OpenCL C or SPIR-V counterparts, you're right.

> === Open Questions ===

Reviewer: Didn't we have anything unsolved?

Author: Added an open question.

// Copyright 2023 The Khronos Group. This work is licensed under a
// Creative Commons Attribution 4.0 International License; see
// http://creativecommons.org/licenses/by/4.0/
= cl_exp_tensor

:source-highlighter: coreray

[[cl_exp_tensor]]
== Tensor Data Type

This extension provides a new opaque OpenCL datatype called
`cl_tensor`. It is used for storing N-dimensional tensor data in
implementation-defined memory layout which may be optimized based on
tensor's use cases. The datatype is designed to be efficiently used
within the `cl_khr_command_buffers` extension to capture task graphs
which can utilize tensors as input, output and temporary storage.

=== General information

> 

Reviewer: Let's write a motivational text here which is interfacing to the DBKs (refer to the WiP spec) on the backend, and to better interface to higher-level models for neural networks and other libraries/formats dealing with tensors.
